### DISCUSSION TITLE
Detect account change

### DISCUSSION THREAD

**Discussion by sjshon** [wevm/wagmi#893] [Q&A]
Does wagmi support account change detection?

Using `useSwitchNetwork` was very useful and thought it'd be nice to have something similar like `useSwitchAccount` 

I've been trying with below approach by switching between accounts in metamask, but I don't see any detection.

```
  import { useProvider } from 'wagmi';

  const AppLayout = ({ children, layoutType }) => {
      const provider = useProvider();

      provider.on("accountsChanged", (accounts: string[]) => {
        console.log(accounts);
      });
  } 
```
What's the suggested way with wagmi?

**Comment by tmm** [comment]
wagmi listens for account changes. You can compose `useAccount`, `useConnect`, and `useDisconnect` in a variety of ways.

**Reply by sjshon** [reply]
ok i've come up with this way
```
 const { address } = useAccount();
 const { disconnectAsync } = useDisconnect();

  useEffect(() => {
    // store connected wallet address to local storage to keep track
    if (address) {
      localStorage.setItem("currentAddress", address);
    }
  }, []);

  useEffect(() => {
    if (address) {
      const currentAddress = localStorage.getItem("currentAddress");
      
      // when current wallet address and switched address is different, disconnect
      if (currentAddress?.toLowerCase() !== address.toLowerCase()) {
        localStorage.setItem("currentAddress", address);
        
        (async function () {
          await disconnectAsync();
        })();
      }
    }
  }, [address]);
```
This works, but any suggestion to avoid local storage?

**Reply by tmm** [reply]
Not sure about your specific use, but I would look at using hook configuration callbacks (e.g. `onConnect` for `useAccount`, `onSuccess` for `useConnect`, etc.) instead of `useEffect`.

Also, wagmi has its own internal storage you can configure when creating the wagmi client. It keeps track of things like the connected account, active chain id, etc.

**Reply by SpasiboKojima** [reply]
> Not sure about your specific use, but I would look at using hook configuration callbacks (e.g. `onConnect` for `useAccount`, `onSuccess` for `useConnect`, etc.) instead of `useEffect`.

But none of them are fired after account change?

**Reply by sjshon** [reply]
coming back to this again @tmm 
```
  useEffect(() => {
     provider.on("accountsChanged", (accounts: string[]) => {
          //....
     });
  });
```
it would be great if above can work. There are some cases where we need that behavior. 
Say when they switched account, we should log them out.

**Reply by tmm** [reply]
You should listen to events on the active `connector`, not `provider`. Check out how this works for the wagmi `Client` [here](https://github.com/wagmi-dev/wagmi/blob/main/packages/core/src/client.ts#L317).

**Reply by sjshon** [reply]
great! that surely helped.

```
  const { connector: activeConnector } = useAccount();
  
  useEffect(() => {
    if (activeConnector) {
      activeConnector.on("change", logout);
    }
  }, [activeConnector]);
```

**Reply by justingreenberg** [reply]
> great! that surely helped.
> 
> ```
>   const { connector: activeConnector } = useAccount();
>   
>   useEffect(() => {
>     if (activeConnector) {
>       activeConnector.on("change", logout);
>     }
>   }, [activeConnector]);
> ```

be sure to return a cleanup function to unbind subscription

**Reply by justingreenberg** [reply]
@kkagill use key of returned data object to determine source of change:

```ts
  const { connector: activeConnector } = useAccount()

  useEffect(() => {
	const handleConnectorUpdate = ({account, chain}: ConnectorData) => {
      if (account) {
        console.log('new account', account)
      } else if (chain) {
        console.log('new chain', chain)
      }
    }

    if (activeConnector) {
      activeConnector.on('change', handleConnectorUpdate)
    }

	return () => activeConnector.off('change', handleConnectorUpdate)
  }, [activeConnector])
```

**Reply by kkagill** [reply]
@justingreenberg thank you! working well
